src/editor/operations.ts

```diff
@@ -32,17 +75,68 @@ export function replaceRangeAndExpandSelection(range: Range, newParts: Part[]):
     });
 }
 
-export function replaceRangeAndMoveCaret(range: Range, newParts: Part[], offset = 0): void {
+export function replaceRangeAndMoveCaret(range: Range, newParts: Part[], offset = 0, atNodeEnd = false): void {
     const { model } = range;
     model.transform(() => {
         const oldLen = range.length;
         const addedLen = range.replace(newParts);
         const firstOffset = range.start.asOffset(model);
-        const lastOffset = firstOffset.add(oldLen + addedLen + offset);
+        const lastOffset = firstOffset.add(oldLen + addedLen + offset, atNodeEnd);
         return lastOffset.asPosition(model);
     });
 }
 
+/**
+ * Replaces a range with formatting or removes existing formatting and
+ * positions the cursor with respect to the prefix and suffix length.
+ * @param {Range} range the previous value
+ * @param {Part[]} newParts the new value
+ * @param {boolean} rangeHasFormatting the new value
+ * @param {number} formatStringLength the length of the format string, assumed to be 0 when not formatted
+ */
+export function replaceRangeAndAutoAdjustCaret(
+    range: Range,
+    newParts: Part[],
+    rangeHasFormatting = false,
+    prefixLength = 0,
+    suffixLength = prefixLength,
+): void {
+    const { model } = range;
+    const lastStartingPosition = range.getLastStartingPosition();
+    const relativeOffset = lastStartingPosition.offset - range.start.offset;
+    const distanceFromEnd = range.length - relativeOffset;
+    // Handle edge case were the caret is located within the suffix or prefix
+    if (rangeHasFormatting) {
+        if (relativeOffset < prefixLength) { // Was the caret at the left format string?
+            replaceRangeAndMoveCaret(range, newParts, -(range.length - 2 * suffixLength));
+            return;
+        }
+        if (distanceFromEnd < suffixLength) { // Was the caret at the right format string?
+            replaceRangeAndMoveCaret(range, newParts, 0, true);
+            return;
+        }
+    }
+    // Calculate new position with respect to the previous position
+    model.transform(() => {
+        const offsetDirection = Math.sign(range.replace(newParts)); // Compensates for shrinkage or expansion
+        const atEnd = distanceFromEnd === suffixLength;
+        return lastStartingPosition.asOffset(model).add(offsetDirection * prefixLength, atEnd).asPosition(model);
+    });
+}
+
+const isSameWord = (_index: number, offset: number, part: Part) => {
+    return part.text[offset] !== " " && part.text[offset] !== "+" && part.type === Type.Plain;
```

Comment:
Is this list of characters sufficient? I don't think punctuation and parentheses should be considered parts of the word, but the specifics are up to discussion.

Reply:
I mean word is used a bit freely here, but I don't really have any better name right now.
It basically boils down to where you want to auto-selection for formatting to stop.
I think one usually wants to include parentheses and punctuation, but I actually excluded these previously, so I am also not 100% sure about this design-wise. Technically, the `+` could be also removed to match GitHub's behavior for example.